### PR TITLE
[SPARK-14558][CORE] In ClosureCleaner, clean the outer pointer if it's a REPL line object

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

When we clean a closure, if its outermost parent is not a closure, we won't clone and clean it as cloning user's objects is dangerous. However, if it's a REPL line object, which may carry a lot of unnecessary references(like hadoop conf, spark conf, etc.), we should clean it as it's not a user object.

This PR improves the check for user's objects to exclude REPL line object.
## How was this patch tested?

existing tests.

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @yhuai @JoshRosen @rxin

File: core/src/main/scala/org/apache/spark/util/ClosureCleaner.scala [hunk 1]
```diff
@@ -77,35 +77,19 @@ private[spark] object ClosureCleaner extends Logging {
    */
   private def getInnerClosureClasses(obj: AnyRef): List[Class[_]] = {
     val seen = Set[Class[_]](obj.getClass)
-    var stack = List[Class[_]](obj.getClass)
+    val stack = Stack[Class[_]](obj.getClass)
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/util/ClosureCleaner.scala]: kind of unrelated, but it's obvious that using `Stach` is more efficient here.

File: core/src/main/scala/org/apache/spark/util/ClosureCleaner.scala [hunk 2]
```diff
@@ -77,35 +77,19 @@ private[spark] object ClosureCleaner extends Logging {
...
       for (cls <- set -- seen) {
         seen += cls
-        stack = cls :: stack
+        stack.push(cls)
       }
     }
     (seen - obj.getClass).toList
   }
 
-  private def createNullValue(cls: Class[_]): AnyRef = {
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/util/ClosureCleaner.scala]: This method is never used.

**Comment by liancheng** [comment]
LGTM

**Comment by liancheng** [comment]
retest this please

**Comment by cloud-fan** [comment]
cc @andrewor14

**Comment by yhuai** [comment]
@JoshRosen Can you take a look?

File: core/src/main/scala/org/apache/spark/util/ClosureCleaner.scala [hunk 3]
```diff
@@ -233,16 +217,22 @@ private[spark] object ClosureCleaner extends Logging {
...
-      // class without cloning it since we don't want to clone the user's objects.
-      // Note that we still need to keep around the outermost object itself because
-      // we need it to clone its child closure later (see below).
-      logDebug(s" + outermost object is not a closure, so do not clone it: ${outerPairs.head}")
-      parent = outerPairs.head._2 // e.g. SparkContext
-      outerPairs = outerPairs.tail
-    } else if (outerPairs.size > 0) {
-      logDebug(s" + outermost object is a closure, so we just keep it: ${outerPairs.head}")
+    if (outerPairs.size > 0) {
+      if (isClosure(outerPairs.head._1)) {
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/util/ClosureCleaner.scala]: not really your code, but can you do:
```
val (outermostClass, outermostObject) = outerPairs.head
if (isClosure(outermostClass)) {
  ...
} else if (outermostClass.getName.startsWith("$line")) {
  ...
} else {
  ...
  parent = outermostObject
  outerPairs = outerPairs.tail
}
```

File: core/src/main/scala/org/apache/spark/util/ClosureCleaner.scala [hunk 4]
```diff
@@ -233,16 +217,22 @@ private[spark] object ClosureCleaner extends Logging {
...
+        logDebug(s" + outermost object is a closure, so we just keep it: ${outerPairs.head}")
+      } else if (outerPairs.head._1.getName.startsWith("$line")) {
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/util/ClosureCleaner.scala]: can you add a short comment in this `if` case describing why we treat REPL classes as a special case (and include `SPARK-14558`)

File: core/src/main/scala/org/apache/spark/util/ClosureCleaner.scala [hunk 5]
```diff
@@ -233,16 +217,22 @@ private[spark] object ClosureCleaner extends Logging {
...
-      // Note that we still need to keep around the outermost object itself because
-      // we need it to clone its child closure later (see below).
-      logDebug(s" + outermost object is not a closure, so do not clone it: ${outerPairs.head}")
-      parent = outerPairs.head._2 // e.g. SparkContext
-      outerPairs = outerPairs.tail
-    } else if (outerPairs.size > 0) {
-      logDebug(s" + outermost object is a closure, so we just keep it: ${outerPairs.head}")
+    if (outerPairs.size > 0) {
+      if (isClosure(outerPairs.head._1)) {
+        logDebug(s" + outermost object is a closure, so we just keep it: ${outerPairs.head}")
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/util/ClosureCleaner.scala]: I wrote this comment originally and now found it confusing. Can you say `so we clone it` instead? Same in L224

**Comment by andrewor14** [comment]
@cloud-fan Can you add a test to `ClosureCleanerSuite2`? Otherwise this LGTM.

File: core/src/main/scala/org/apache/spark/util/ClosureCleaner.scala (hunk 5 above)
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/util/ClosureCleaner.scala]: +1; I also was confused by this comment's existing wording and it look me a while to understand that "keeping it" here means making it eligible for cloning and cleaning further down.

File: core/src/main/scala/org/apache/spark/util/ClosureCleaner.scala [hunk 6]
```diff
@@ -20,6 +20,7 @@ package org.apache.spark.util
 import java.io.{ByteArrayInputStream, ByteArrayOutputStream}
 
 import scala.collection.mutable.{Map, Set, Stack}
+import scala.language.existentials
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/util/ClosureCleaner.scala]: not used

File: core/src/main/scala/org/apache/spark/util/ClosureCleaner.scala [hunk 7]
```diff
@@ -19,7 +19,8 @@ package org.apache.spark.util
 
 import java.io.{ByteArrayInputStream, ByteArrayOutputStream}
 
-import scala.collection.mutable.{Map, Set}
+import scala.collection.mutable.{Map, Set, Stack}
+import scala.language.existentials
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/util/ClosureCleaner.scala]: i'll remove this

**Comment by andrewor14** [comment]
LGTM merging into master thanks

File: core/src/main/scala/org/apache/spark/util/ClosureCleaner.scala (hunk 7 above)
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/util/ClosureCleaner.scala]: oops, I forgot to. Oh well

Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/util/ClosureCleaner.scala]: Actually it's needed, or you will get compile warning, it's caused by this line: `val (outermostClass, outermostObject) = outerPairs.head`

Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/util/ClosureCleaner.scala]: oh, good thing I forgot!

**Comment by a-roberts** [comment]
Hi, with this new test I'm seeing large deviations using IBM JDKs and different platforms, for example:

`java.lang.AssertionError: assertion failed: deviation too large: 0.25359712230215825, first size: 26688, second size: 19920`

Is the 20% deviation particularly important? Why this number?

FYI here's the comparison between different Java vendors and architectures:
- OpenJDK on Intel (passes), cacheSize1: 180392 and cacheSize2: 187896 (4.07%)
- IBM JDK with SUSE on zSystems (fails): cacheSize1: 26688 and cacheSize2: 19920 (29%)
- IBM JDK with Ubuntu 14 04 on Power 8 LE (fails): cacheSize1: 26688 and cacheSize2: 19920 (29%)
- IBM JDK with Ubuntu 14 04 on Intel (fails): cacheSize1: 354692 and cacheSize2: 263800 (29.3%)

I'll look into this, interesting that with IBM Java it's always a very similar and much larger percentage

**Comment by cloud-fan** [comment]
@a-roberts wow that's interesting, I think maybe IBM JDK brings more information to scala REPL line object, which increases the cache size. It will be great if you can look into it, thanks!

BTW if this problem do matters for you, feel free to send a PR to increase the threshold(20%).

**Comment by a-roberts** [comment]
@cloud-fan I've had a closer look at this and think a more robust method would be to use weak references to identify when an object is out of scope, with IBM Java we see the 29% reduction between cache size 1 and cache size 2 but with OpenJDK we see a 4% increase, suggesting that we can't rely on the sizes being similar across JDK vendors, now thinking this is a test case issue rather than a problem in the ClosureCleaner or IBM Java code.

With IBM Java our second cache size (after repartitioning) is much smaller; repartitioning uses ContextCleaner whereas with OpenJDK it grows. Either we have a bigger memory footprint or the cached size is being calculated incorrectly (looks fine to me and we actually have smaller object sizes). The problem on Z was due to using repl/pom.xml instead of pom.xml in the Spark home directory (same result if we use the right pom.xml file) so can be discarded for this discussion.

I'm going to figure out what's in the Scala REPL line objects between vendors, I think the intention of this commit is to test that the REPL line object is being cleaned but the assertion in place at the moment doesn't look to be correct (the size is bigger after the cleaning and cacheSize2 is the result of cleaning if I'm understanding the code correctly), have I missed a trick?

**Comment by cloud-fan** [comment]
`cacheSize1` and `cacheSize2` are both the size after cleaning. The difference is that, `cacheSize1` is the size after cleaned the data with line object reference, `cacheSize2` is the size after cleaned the data without line object reference.

And yes, the test missed the difference of size of Scala REPL line objects between vendors, feel free to send a PR to fix it and thanks for investigating this!

**Comment by a-roberts** [comment]
> cacheSize1 and cacheSize2 are both the size after cleaning. The difference is that, cacheSize1 is the size after cleaned the data with line object reference, cacheSize2 is the size after cleaned the data without line object reference.

Looking for clarity here, is it true that clean "with the reference" should be bigger (cacheSize1) and clean "without the reference" should be smaller (cacheSize2)? 

OpenJDK, cacheSize1: 180392, cacheSize2: 187896 (bigger without the line object reference)

IBM JDK, cacheSize1: 354692, cacheSize2: 263800 (smaller without the line object reference)

What exactly does "without line object reference" mean and should cacheSize1 be smaller or bigger than cacheSize2?

I know the SizeEstimator overestimates for IBM Java so our cached footprint is much larger (handling this), so because of the larger difference we get this test failing, OpenJDK **fails** with Kryo and IBM **passes** with Kryo for this test.

A better check would be to run with and without the closure cleaner change and to check the second result is less by the size of the line object, so based on our cacheSize2 being smaller (without the line object reference), I'm thinking that IBM Java functions as expected and OpenJDK doesn't - but this depends on my questions above, interested to hear what you think @cloud-fan

**Comment by cloud-fan** [comment]
`cacheSize1` and `cacheSize2` should be almost same, because they cache the same data, with different RDD partition numbers. However, before this PR, `cacheSize1` is much larger because it references line object and doesn't clean it.

> A better check would be to run with and without the closure cleaner change
Yea, this is what I did locally, but how to write a test for it?

**Pull request closed without merging**